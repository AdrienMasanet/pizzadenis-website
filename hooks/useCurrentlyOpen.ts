import businessDays from "@/data/businessDays";
import BusinessDay from "@/interfaces/BusinessDay";
import formatToHumanHour from "@/utils/formatToHumanHour";

/**
 * The status of the business at the time the hook is called, depending on the openingHours array in data/openingHours.ts.
 * @param {Date} today The date to check the status of the business. Defaults to the current date. Useful for testing !
 * @returns {[boolean, string]} An array containing the open status and a message.
 */
const useCurrentlyOpen = (today: Date = new Date()) => {
  const currentHour = parseFloat(today.getHours() + "." + today.getMinutes());
  const currentDay: BusinessDay = businessDays[today.getDay()];
  const isOpen = currentHour >= currentDay.startHour && currentHour < currentDay.endHour;
  let message = "";

  const getNextOpeningDay = (from: number, tomorrow: boolean = true): BusinessDay => {
    let dayIterator = from + 1;

    if (dayIterator > 6) {
      dayIterator = 0;
    }

    // If the next day is a day off, recursively call the function to get the following day with tomorrow set to false.
    if (businessDays[dayIterator].dayOff) {
      return getNextOpeningDay(dayIterator, false);
    } else {
      // If the next opening day is tomorrow, we return the day but with "demain" as the name of the day.
      if (tomorrow) return { ...businessDays[dayIterator], name: "demain" };

      return businessDays[dayIterator];
    }
  };

  const getMinutesDifference = (state: "closing" | "opening"): number => {
    const eventTime: Date = new Date(`${today.getFullYear()}-${today.getMonth() + 1}-${today.getDate()} ${state === "closing" ? currentDay.endHour : currentDay.startHour}:00:00`);
    const millisecondsDifference = (eventTime.getTime() - today.getTime()) as number;
    const minutesDifference = Math.floor(millisecondsDifference / 60000);

    return minutesDifference;
  };

  if (getMinutesDifference("opening") > 90) {
    if (!currentDay.dayOff && currentDay.startHour) {
      message = `Nous ouvrons à ${formatToHumanHour(currentDay.startHour)}h !`;
    } else {
      message = "Nous sommes fermés aujourd'hui...";
    }
  } else if (getMinutesDifference("opening") <= 90 && getMinutesDifference("opening") > 30) {
    // Between 90 and 30 minutes before opening
    message = "Nous ouvrons bientôt !";
  } else if (getMinutesDifference("opening") <= 30 && getMinutesDifference("opening") > 0) {
    // Between 30 and 0 minutes before opening
    message = "Nous ouvrons très bientôt !";
  } else if (getMinutesDifference("opening") <= 0 && getMinutesDifference("closing") > 60) {
    // Between opening hour and 60 minutes before closing
    message = "Nous sommes ouverts !";
  } else if (getMinutesDifference("closing") <= 60 && getMinutesDifference("closing") > 30) {
    // Between 60 and 30 minutes before closing
    message = `Nous fermons bientôt (dans ${getMinutesDifference("closing")} minutes) !`;
  } else if (getMinutesDifference("closing") <= 30 && getMinutesDifference("closing") > 0) {
    // Between 30 and 0 minutes before closing
    message = `Attention, nous fermons très bientôt (dans ${getMinutesDifference("closing")} minutes) !`;
  } else if (getMinutesDifference("closing") <= 0) {
    message = `Nous sommes fermés... Rendez-vous dès ${getNextOpeningDay(today.getDay()).name} à ${formatToHumanHour(getNextOpeningDay(today.getDay()).startHour)}h !`;
  }

  return [isOpen, message];
};

export default useCurrentlyOpen;
