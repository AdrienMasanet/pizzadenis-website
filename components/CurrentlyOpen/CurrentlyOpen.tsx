import useCurrentlyOpen from "@/hooks/useCurrentlyOpen";
import { useEffect, useState } from "react";

const CurrentlyOpen = () => {
  const [currentDateTime, setCurrentDateTime] = useState(new Date());
  const [isOpen, message] = useCurrentlyOpen(currentDateTime);

  useEffect(() => {
    const interval = setInterval(() => {
      setCurrentDateTime(new Date());
    }, 1000 * 10); // Update opening status every 20 seconds to accurately display minutes left

    return () => clearInterval(interval);
  }, []);

  return (
    <div className="mx-12 md:mx-5">
      <div className={`${isOpen ? "neon-borders" : ""} w-fit px-3 py-2 mx-auto my-10 rounded-lg bg-stone-50/60 border border-black/20 backdrop-blur-sm shadow-lg`}>
        <p>{message}</p>
      </div>
    </div>
  );
};

export default CurrentlyOpen;
