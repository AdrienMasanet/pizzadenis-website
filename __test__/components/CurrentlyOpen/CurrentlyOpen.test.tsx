import { render, screen, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import CurrentlyOpen from "@/components/CurrentlyOpen/CurrentlyOpen";
import useCurrentlyOpen from "@/hooks/useCurrentlyOpen";

// Mock the useCurrentlyOpen hook
jest.mock("@/hooks/useCurrentlyOpen");
const mockedUseCurrentlyOpen = useCurrentlyOpen as jest.Mock;

describe("components | CurrentlyOpen", () => {
  beforeEach(() => {
    // Clear all instances and calls to the mock
    mockedUseCurrentlyOpen.mockClear();
  });

  it("should render the component with the correct message", () => {
    // Set up the mock implementation for the hook
    mockedUseCurrentlyOpen.mockImplementation(() => [true, "Currently open!"]);

    const renderedComponent = render(<CurrentlyOpen />);

    expect(screen.getByText("Currently open!")).toBeInTheDocument();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("should update the opening status and message every 10 seconds", () => {
    // Use fake timers to control the setInterval function
    jest.useFakeTimers();

    mockedUseCurrentlyOpen.mockImplementation(() => [true, "Currently open!"]);

    render(<CurrentlyOpen />);

    // Move the timer forward by 20 seconds
    act(() => {
      jest.advanceTimersByTime(1000 * 10);
    });

    // Check if the useCurrentlyOpen hook was called twice (initial render and after 20 seconds)
    expect(mockedUseCurrentlyOpen).toHaveBeenCalledTimes(2);

    // Restore the real timers
    jest.useRealTimers();
  });

  it("should apply the 'neon-borders' class when the store is open", () => {
    mockedUseCurrentlyOpen.mockImplementation(() => [true, "Currently open!"]);

    render(<CurrentlyOpen />);

    const divElement = screen.getByTestId("container");
    expect(divElement).toHaveClass("neon-borders");
  });

  it("should not apply the 'neon-borders' class when the store is closed", () => {
    mockedUseCurrentlyOpen.mockImplementation(() => [false, "Currently closed!"]);

    render(<CurrentlyOpen />);

    const divElement = screen.getByTestId("container");
    expect(divElement).not.toHaveClass("neon-borders");
  });
});
