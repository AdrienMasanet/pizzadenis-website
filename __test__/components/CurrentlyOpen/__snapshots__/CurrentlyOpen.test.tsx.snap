// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | CurrentlyOpen should render the component with the correct message 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mx-12 md:mx-5"
      >
        <div
          class="neon-borders w-fit px-3 py-2 mx-auto my-10 rounded-lg bg-stone-50/60 border border-black/20 backdrop-blur-sm shadow-lg"
          data-testid="container"
        >
          <p>
            Currently open!
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mx-12 md:mx-5"
    >
      <div
        class="neon-borders w-fit px-3 py-2 mx-auto my-10 rounded-lg bg-stone-50/60 border border-black/20 backdrop-blur-sm shadow-lg"
        data-testid="container"
      >
        <p>
          Currently open!
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
